grading_bin/insert_cells.py

```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/python3
+"""
+Script to insert missing checkpoint cells from an instructor notebook into a student's
+submitted notebook.
+
+This script compares the cells in both notebooks based on their IDs and inserts any missing
+cells from the instructor notebook into the student notebook.
+
+Student cells are preserved if they exist in both notebooks. Extra student cells
+missing from the instructor notebook are not added to the final output notebook for
+consistency.
+"""
+
+import argparse
+# pylint: disable=import-error
+import nbformat
+
+
+def insert_cells(instructor_path, student_path, output_path):
+    """
+    Insert missing cells from the instructor notebook into the student notebook.
+    The function reads both notebooks, compares their cells by ID, and inserts any missing
+    cells from the instructor notebook into the student notebook.
+    Args:
+        instructor_path (str): Path to the instructor notebook.
+        student_path (str): Path to the student's submitted notebook.
+        output_path (str): Path where the final notebook will be saved.
+    """
+    # Load instructor and student notebooks
+    try:
+        with open(instructor_path, 'r', encoding='utf-8') as f:
+            instructor_nb = nbformat.read(f, as_version=nbformat.NO_CONVERT)
+        print(f"Loaded instructor notebook from '{instructor_path}'")
+        with open(student_path, 'r', encoding='utf-8') as f:
+            student_nb = nbformat.read(f, as_version=nbformat.NO_CONVERT)
+        print(f"Loaded student notebook from '{student_path}'")
+    except FileNotFoundError as e:
+        print(f"Error: {e}")
+        return
+
+    student_ids = {cell.id: cell for cell in student_nb.cells if 'id' in cell}
+
+    new_cells = []
+    inserted_count = 0
+
+    print("Inserting missing checkpoint cells from instructor notebook into student notebook...")
+    for cell in instructor_nb.cells:
+        # Compares cell IDs to see if the cell exists in the student notebook
+        if 'id' in cell and cell.id in student_ids:
+            # If the cell exists in the student notebook, keep it
```

Comment:
We might want to add an argument for which should win if there is a conflict, sometimes an instructor might want to let there be a student provided section and sometimes the instructor might want to make sure that their code is what is being graded.